orchagent/mirrororch.cpp

```diff
@@ -613,6 +671,83 @@ bool MirrorOrch::updateSession(const string& name, MirrorEntry& session)
     return ret;
 }
 
+bool MirrorOrch::setUnsetPortMirror(Port port,
+                                    bool ingress,
+                                    bool set,
+                                    sai_object_id_t sessionId)
+{
+    sai_status_t status;
+    sai_attribute_t port_attr;
+    port_attr.id = ingress ? SAI_PORT_ATTR_INGRESS_MIRROR_SESSION:
+                           SAI_PORT_ATTR_EGRESS_MIRROR_SESSION;
+    if (set)
+    {
+        port_attr.value.objlist.count = 1;
+        port_attr.value.objlist.list = (long unsigned int *)calloc(port_attr.value.objlist.count, sizeof(sai_object_id_t));
+        port_attr.value.objlist.list[0] = sessionId;
+    }
+    else
+    {
+        port_attr.value.objlist.count = 0;
+    }
+
+    if (port.m_type == Port::LAG)
+    {
+        vector<Port> portv;
+        m_portsOrch->getLagMember(port, portv);
+        for (const auto p : portv)
+        {
+            status = sai_port_api->set_port_attribute(p.m_port_id, &port_attr);
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("Failed to configure mirror session port %s: %s, status %d, sessionId %x\n",
+                                port.m_alias.c_str(), p.m_alias.c_str(), status, sessionId);
+                return false;
+            }
+        }
+    }
+    else if (port.m_type == Port::PHY)
+    {
+        status = sai_port_api->set_port_attribute(port.m_port_id, &port_attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to configure mirror session to port %s, status %d, sessionId %x\n",
+                            port.m_alias.c_str(), status, sessionId);
+            return false;
+        }
+    }
+    return true;
+}
+
+bool MirrorOrch::configurePortMirrorSession(const string& name, MirrorEntry& session, bool set)
+{
+    auto ports = tokenize(session.src_port, ',');
+    if (ports.size() != 0)
+    {
+        for (auto alias : ports)
+        {
+            Port port;
```

Comment:
should we continue or return failure when one of the port/lag mirror session set fails?

Reply:
Currently it returns when one of the config fails from SAI. The failures from SAI are critical and orchagent terminate is initiated from syncd. I am not sure if continue will help. Please suggest.

Reply:
There will be inconsistency for some cases, 

- Mirror on both the directions, when port ingress mirror update succeeds and port egress mirror update fails, there will be inconsistency in the config DB and asic DB.
- Mirror on LAG, when mirror session set/unset fails on some members, there will be inconsistency on some LAG members. But the actual config is for LAG and not the members.

We could go by a common model. 

@lguohan Any thoughts??

Reply:
I see similar behaviour in  setHostIntfsStripTag in orchagent/portsorch.cpp 
When the port is LAG and the SAI API (set_hostif_attribute) fails for any port in LAG, 
then the function returns failure to the caller.

Reply:
Yeah, we have similar inconsistency in other modules as well.

Reply:
@lguohan @prsunny do you have thoughts for this?

Reply:
This would be tricky. The actual return value of the function is not checked. One suggestion is, if there is a failure, you can invoke the deactivate session and exit cleanly. Or you can log the error and expect user to delete the session.  Have you checked the flow for "unsetting" session on a port which is not originally "set"?

Reply:
Sorry I didn't get which function return value is not handled here. 
configurePortMirrorSession return value is already handled and session becomes inactive when some config fails on any port, and user can delete the session.
Setting a port which is already set gives error SAI_STATUS_ITEM_ALREADY_EXISTS
Unsetting a port which is not set doesnt give any error.

We are already logging error and notifying user about failure.  As i mentioned earlier, there are other codes which follow similar approach as we are doing here.

Please let me know if u have strong preference for calling deactivate session on failure. 
We need to handle both set and unset differently, I will take care in next PR.

Reply:
So typically, there is a retry logic in `orchagent ` that if something fails, it keeps retrying (if you return `false` to `doTask`). What I meant was, this function returns to `activateSession `and subsequent return value is ignored. So essentially, if something fails, there is no retry. Agree with the current approach, you can enhance it in next PR.

Reply:
Thank you.